File: aula_6/aula_6.py
from math import exp
import matplotlib.pyplot as plt
import numpy as np

# aula 6 - algoritmo do método da bissecção e algumas de suas variações


# plotar gráfico da função
def f(x):
    return exp(x) - x - 2


xx = np.arange(-5, 5, 0.1)
yy = list(map(f, xx))
fig, ax = plt.subplots()
ax.plot(xx, yy)
ax.grid()
# plt.show()


def bis(func, a, b, n, tol):
    """
        método da bissecção
    """
    fa = func(a)
    x0 = a

    for i in range(n):
        x = a + (b - a) / 2
        fx = func(x)

        print("{} {:1.5e} {:1.2e}".format(i, x, abs(x-x0)))

        if fx == 0 or abs(x-x0) < tol:
            return x

        x0 = x

        if fa * fx < 0:
            b = x
        else:
            a = x
            fa = func(a)
    print("num. máximo de iterações foi atingido")
    return x


def tris(func, a, b, n, tol):
    """
        método da bissecção utilizando 3 segmentos
    """
    fa = func(a)
    x0 = a

    for i in range(n):
        x = a + (b - a) / 3
        fx = func(x)

        print("{} {:1.5e} {:1.2e}".format(i, x, abs(x-x0)))

        if fx == 0 or abs(x-x0) < tol:
            return x

        x0 = x

        if fa * fx < 0:
            b = x
        else:
            a = x
            fa = func(a)
            x = a + (b-a) / 2
            fx = func(x)

            if fa * fx < 0:
                b = x
            else:
                a = x
                fa = func(a)

    print("num. máximo de iterações foi atingido")
    return x


def secs(func, a, b, n, tol):
    """
        método da bissecção utilizando reta secante
    """
    fa = func(a)
    x0 = a

    for i in range(n):
        x = a - fa * (b-a)/(func(b) - fa)
        fx = func(x)

        print("{} {:1.5e} {:1.2e}".format(i, x, abs(x-x0)))

        if fx == 0 or abs(x-x0) < tol:
            return x

        x0 = x

        if fa * fx < 0:
            b = x
        else:
            a = x
            fa = func(a)
    print("num. máximo de iterações foi atingido")
    return x

File: aula_6/aula_6.m

#Aula 6 - algoritmo do metodo da bissecçao e algumas de suas variaçoes

#Plotar grafico da funçao
f = @(x) exp(x)-x-2;
xx = linspace(1,2);
#plot(xx, f(xx))

#Metodo da bissecçao
function x = bis(f, a, b, N, TOL)
  i = 1;
  fa = f(a);
  x0 = a;

  while (i <= N)
     x = a + (b - a) / 2;
     fx = f(x);

     printf("%i %1.5e %1.2e\n", i, x, abs(x-x0))

     if ((fx == 0 ) || (abs(x-x0) < TOL))
      return;
     endif

     i += 1;
     x0 = x;

     if (fa * fx < 0)
      b = x;
     else
      a = x;
      fa = f(a);
     endif

  endwhile
  error("num. maximo de iteraçoes foi atingido")
endfunction

#Metodo da bissecçao utilizando 3 segmentos
function x = tris(f, a, b, N, TOL)
  i = 1;
  fa = f(a);
  x0 = a;

  while (i <= N)
     x = a + (b - a) / 3;
     fx = f(x);

     printf("%i %1.5e %1.2e\n", i, x, abs(x-x0))

     if ((fx == 0 ) || (abs(x-x0) < TOL))
      return;
     endif

     i += 1;
     x0 = x;

     if (fa * fx < 0)
      b = x;
     else
      a = x;
      fa = f(a);
      x = a + (b-a)/2;
      fx = f(x);
      if (fa * fx < 0)
        b = x;
      else
        a = x;
        fa = f(a);
      endif
     endif

  endwhile
  error("num. maximo de iteraçoes foi atingido")
endfunction

#Metodo da bissecçao utilizando reta secante
function x = secs(f, a, b, N, TOL)
  i = 1;
  fa = f(a);
  x0 = a;

  while (i <= N)
     x = a - fa*(b-a)/(f(b)-fa);
     fx = f(x);

     printf("%i %1.5e %1.2e\n", i, x, abs(x-x0))

     if ((fx == 0 ) || (abs(x-x0) < TOL))
      return;
     endif

     i += 1;
     x0 = x;

     if (fa * fx < 0)
      b = x;
     else
      a = x;
      fa = f(a);
     endif

  endwhile
  error("num. maximo de iteraçoes foi atingido")
endfunction
